connect-location/src/main/java/com/ifttt/location/BackupGeofenceMonitor.java

```diff
@@ -0,0 +1,234 @@
+     * cannot be found.
+     */
+    @Nullable
+    MonitoredGeofence.GeofenceState getState(String fenceKey) {
+        Map<String, MonitoredGeofence> map = getMonitoredGeofences();
+        MonitoredGeofence currentGeofence = map.get(fenceKey);
+        return currentGeofence != null ? currentGeofence.state : null;
+    }
+
+    /**
+     * @return all monitored geofence states.
+     */
+    Map<String, MonitoredGeofence> getMonitoredGeofences() {
+        String monitoredGeofencesString = sharedPreferences.getString(PREF_KEY_MONITORED_GEOFENCES, null);
+        if (monitoredGeofencesString == null) {
+            return Collections.emptyMap();
+        }
+
+        try {
+            return jsonAdapter.fromJson(monitoredGeofencesString);
+        } catch (IOException e) {
+            return Collections.emptyMap();
+        }
+    }
+
+    void checkMonitoredGeofences(double lat, double lng, OnEventUploadListener listener) {
+        Map<String, MonitoredGeofence> monitoredGeofences = getMonitoredGeofences();
+        for (Map.Entry<String, MonitoredGeofence> entry : monitoredGeofences.entrySet()) {
+            MonitoredGeofence monitoredGeofence = entry.getValue();
+            requireNonNull(monitoredGeofence.value.radius);
+
+            Location newLocation = new Location(LocationManager.PASSIVE_PROVIDER);
+            newLocation.setLatitude(lat);
+            newLocation.setLongitude(lng);
+
+            Location geofence = new Location(LocationManager.PASSIVE_PROVIDER);
+            geofence.setLatitude(monitoredGeofence.value.lat);
+            geofence.setLongitude(monitoredGeofence.value.lng);
+
+            MonitoredGeofence.GeofenceState newState;
+            LocationEventUploader.EventType newType;
+            if (newLocation.distanceTo(geofence) >= monitoredGeofence.value.radius) {
+                newState = Exited;
+                newType = Exit;
+            } else {
+                newState = Entered;
+                newType = Entry;
+            }
+
+            boolean isMatchingEntryEvent = newState == Entered && monitoredGeofence.type == Entry;
+            boolean isMatchingExitEvent = newState == Exited && monitoredGeofence.type == Exit;
+
+            if (newState != monitoredGeofence.state) {
+                if (monitoredGeofence.state != Init && (isMatchingEntryEvent || isMatchingExitEvent)) {
+                    listener.onUploadEvent(entry.getKey(), newType);
+                } else {
+                    String fenceKey = entry.getKey();
+                    listener.onUploadSkipped(
+                        fenceKey,
+                            "Upload is skipped, with state: "
+                            + monitoredGeofence.state
+                            + ", type: "
+                            + monitoredGeofence.type
+                            + ", event: "
+                            + newState
+                    );
+                }
```

Comment:
The `else` code block will be executed for duplicates right? Can we add a comment here for that?